fix Tuist source glob not handling nested folder symlinks

LorDisturbia opened the pull request:
Resolves https://github.com/tuist/tuist/issues/3268
Request for comments document (if applies):

### Short description 📝

This PR enhances the FileManager with a version of `subpaths(atPath:)` that correctly follows symlinks.

### Checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase.
- [x] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [x] ~~In case the PR introduces changes that affect users, the documentation has been updated.~~

adellibovi commented:
Could we do a performance test to check what's the impact of this? I remember glob could be quite sensitive.
We can use the fixturegen to create few projects with ~1000 sources files and see what's the impact on performance with this change.

LorDisturbia commented:
@adellibovi Sure! Can you elaborate a bit on how you'd proceed with the testing? I don't have experience with fixtures

LorDisturbia commented:
Ok here's what I did:

- I generated 5 projects, each with 5 targets, each with 1000 files (using `fixturegen --sources 1000`)
- I checked out the latest commit on my branch, built it, and ran
```sh
for i in {1..5}; do (cd "Project$i" && ../../.build/x86_64-apple-macosx/debug/tuist generate); done
```
- I collected the times given by Tuist (`Total time taken: X.XXXs`)
- I calculated the mean and standard deviation
- I repeated all steps above after checking out the `main` branch (of course, deleting and recreating the projects to prevent any kind of caching)

The results are:
```
With my change:
=> Mean 12.677s, standard deviation: 1.89s

Before my change:
=> Mean 12.575s, standard deviation: 1.11s
```

I don't see any statistically significant difference

adellibovi commented:
@LorDisturbia for the future, so that you are aware, we have a command to do so under `project/tuistbench`, more info here: https://docs.tuist.io/contributors/performance-testing 😄 
Anyway, what you did is quite correct, but I see you used the debug build. We should use release builds for benchmark testing, mind doing that? Thanks 🙏

adellibovi commented:
Ah sorry @LorDisturbia, it is better to use the `tuistbench`, it takes care of cold and warm generation as well

LorDisturbia commented:
Hey there! I ended up running `./fourier benchmark` and these are the results:

| Fixture         | New    | Old  | Delta    |
| --------------- | ------ | ---- | -------- |
| 50_projects _(cold)_ | 4,52s | 4,40s | ⬆︎ +0,12s +2,70% |
| 50_projects _(warm)_ | 2,43s | 2,36s | ⬆︎ +0,07s +2,89% |
| 2000_sources _(cold)_ | 8,95s | 8,62s | ⬆︎ +0,33s +3,82% |
| 2000_sources _(warm)_ | 8,57s | 8,16s | ⬆︎ +0,41s +5,01% |
| ios_app_with_static_frameworks _(cold)_ | 1,09s | 1,11s | ⬇︎ -0,02s -2,00% |
| ios_app_with_static_frameworks _(warm)_ | 0,16s | 0,21s | ⬇︎ -0,05s -22,26% |
| ios_app_with_framework_and_resources _(cold)_ | 1,18s | 1,21s | ⬇︎ -0,03s -2,56% |
| ios_app_with_framework_and_resources _(warm)_ | 0,27s | 0,27s | ≈ |
| ios_app_with_transitive_framework _(cold)_ | 0,82s | 0,83s | ≈ |
| ios_app_with_transitive_framework _(warm)_ | 0,14s | 0,20s | ⬇︎ -0,07s -32,56% |
| ios_app_with_xcframeworks _(cold)_ | 0,41s | 0,40s | ≈ |
| ios_app_with_xcframeworks _(warm)_ | 0,13s | 0,15s | ≈ |

adellibovi commented:
Seems good, thanks for checking ☺️

adellibovi commented:
@LorDisturbia There seems to be a bug though, I tried this PR on my project and it does not recognize most of the files, basically it only gets the one at the root and not in the subdirectories.

LorDisturbia commented:
I fixed it and added a test. The problem was that my method was returning paths INCLUDING the base path, while the caller expected only subpaths (relative to the base path) just like the system `subpaths(atPath:)` does. So when in `Glob.swift` the caller tried to manually add the base path again it ended up with something that was `/path/to/base/path/to/base/myFolder` instead of /path/to/base/myFolder`, which obviously doesn't exist and is then discarded.

I also fixed the tests and added an extra one just to be safe. The benchmark hasn't changed because the logic inside the filemanager does exactly the same operations.

adellibovi commented:
Hi @LorDisturbia,
The benchmark generates a fixture that only has sources in a directory and not different directories, that is the reason why it seems like there is no impact.
I tried locally with a different fixture that has folders as well and it got +12/14% slower.
Do you mind profiling your code? Maybe there are some things we can improve, I would imagine string interpolation can affect.
In the meanwhile I will update the fixturegen to generate subdirectories as well.

LorDisturbia commented:
I went over a few iterations and ended up with the current solution. The semantics of the method is different, which allowed me to simplify the caller side and avoid N extra calls to isDirectory. This compensates for the cost of resolving symlinks, giving us a net positive change:

| Fixture         | New    | Old  | Delta    |
| --------------- | ------ | ---- | -------- |
| 100_projects _(cold)_ | 2,92s | 3,06s | ⬇︎ -0,14s -4,79% |
| 100_projects _(warm)_ | 2,46s | 2,53s | ⬇︎ -0,07s -2,85% |

LorDisturbia commented:
I also rerun the old benchmarks just to be sure and honestly I don't see anything statistically significant 

| Fixture         | New    | Old  | Delta    |
| --------------- | ------ | ---- | -------- |
| 50_projects _(cold)_ | 4,71s | 4,62s | ⬆︎ +0,09s +1,98% |
| 50_projects _(warm)_ | 2,48s | 2,40s | ⬆︎ +0,08s +3,27% |
| 2000_sources _(cold)_ | 8,96s | 8,92s | ⬆︎ +0,04s +0,46% |
| 2000_sources _(warm)_ | 8,35s | 8,33s | ≈ |
| ios_app_with_static_frameworks _(cold)_ | 1,03s | 1,01s | ≈ |
| ios_app_with_static_frameworks _(warm)_ | 0,13s | 0,13s | ≈ |
| ios_app_with_framework_and_resources _(cold)_ | 1,13s | 1,13s | ≈ |
| ios_app_with_framework_and_resources _(warm)_ | 0,27s | 0,27s | ≈ |
| ios_app_with_transitive_framework _(cold)_ | 0,74s | 0,76s | ≈ |
| ios_app_with_transitive_framework _(warm)_ | 0,13s | 0,13s | ≈ |
| ios_app_with_xcframeworks _(cold)_ | 0,33s | 0,40s | ⬇︎ -0,07s -16,88% |
| ios_app_with_xcframeworks _(warm)_ | 0,13s | 0,13s | ≈ |

adellibovi reviewed (approved):
Thanks @LorDisturbia for resolving the hit on performance 😄

LorDisturbia commented:
No probs! Thanks for pointing them out!

danyf90 commented:
@all-contributors add @LorDisturbia for code
